### PR TITLE
[SPARK-18465] Add 'IF EXISTS' clause to 'UNCACHE' to not throw exceptions when table doesn't exist

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

While this behavior is debatable, consider the following use case:
```sql
UNCACHE TABLE foo;
CACHE TABLE foo AS 
SELECT * FROM bar
```
The command above fails the first time you run it. But I want to run the command above over and over again, and I don't want to change my code just for the first run of it.
The issue is that subsequent `CACHE TABLE` commands do not overwrite the existing table.

Now we can do:
```sql
UNCACHE TABLE IF EXISTS foo;
CACHE TABLE foo AS 
SELECT * FROM bar
```

## How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by brkyvz** [comment]
cc @gatorsmile I changed a test you added. Do you have any strong feelings on this?

**Comment by gatorsmile** [comment]
After fixed the styles, LGTM pending testing.

**Comment by brkyvz** [comment]
@gatorsmile Talking offline with several people, I may put this PR on hold for now since it is  a behavior change. I guess it would be better to go with Options 1 or 2 that I defined in the PR description.

**Comment by gatorsmile** [comment]
@brkyvz Why not supporting both? : )

**Comment by dongjoon-hyun** [comment]
Hi, @brkyvz and @gatorsmile .
Does this proceed to Option 1 or 2 for now? Or is this holding on for next month?

**Comment by brkyvz** [comment]
On hold on my side. Will try to get back to it

**Comment by dongjoon-hyun** [comment]
I see. Thank you for informing that, @brkyvz

**Comment by andrewor14** [comment]
I personally think `UNCACHE TABLE IF EXISTS` is best. It preserves the old behavior but lets the user make sure a table is not cached if they really want.

**Comment by brkyvz** [comment]
Hey @andrewor14 ! I went with your way. @hvanhovell can you take a quick look please? I would really like this to be available in Spark 2.1 (even though it is a new API)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala]: Minor: Lets not log it, logs typically get swallowed by the environment anyway. See for example: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala#L206

**Review by hvanhovell — CHANGES_REQUESTED** [review]
One small comment, otherwise LGTM

**Comment by brkyvz** [comment]
@hvanhovell Done! Thanks for the quick review!

**Comment by hvanhovell** [comment]
LGTM pending jenkins

**Comment by gatorsmile** [comment]
When the target is a view, it becomes a little bit more complex. If this PR does not handle it, we can do it in a separate PR.
```SQL
UNCACHE TABLE IF EXISTS viewName;
```

**Comment by hvanhovell** [comment]
@gatorsmile could you elaborate a little? What am I missing?

**Comment by hvanhovell** [comment]
Merging to master/2.1. Thanks!

**Comment by gatorsmile** [comment]
```Scala
      spark.range(1, 10).toDF("id1").write.format("json").saveAsTable("jt1")
      spark.range(1, 10).toDF("id2").write.format("json").saveAsTable("jt2")
      sql("CREATE VIEW testView AS SELECT * FROM jt1 JOIN jt2 ON id1 == id2")
      // Cache is empty at the beginning
      assert(spark.sharedState.cacheManager.isEmpty)
      sql("CACHE TABLE testView")
      assert(spark.catalog.isCached("testView"))
      // Cache is not empty
      assert(!spark.sharedState.cacheManager.isEmpty)
      // drop a table referenced by a cached view
      sql("DROP TABLE jt1")

-- So far everything is fine

      // Failed to unache the view
      val e = intercept[AnalysisException] {
        sql("UNCACHE TABLE testView")
      }.getMessage
      assert(e.contains("Table or view not found: `default`.`jt1`"))

      // We are unable to drop it from the cache
      assert(!spark.sharedState.cacheManager.isEmpty)
```

@hvanhovell Above is the example.

**Comment by hvanhovell** [comment]
Thanks! Isn't this an existing (separate) problem, that we should fix?

**Comment by gatorsmile** [comment]
Yeah! That is just a relavant problem. Found it when I reviewed another PR: https://github.com/apache/spark/pull/15682

**Comment by hvanhovell** [comment]
I agree that it is relevant, and that we should fix it for 2.1.

**Pull request closed without merging**
